### PR TITLE
Introduce a simple validation hook

### PR DESCRIPTION
**Pull request by jcrossley3**
Partially fulfills #7

SBOM's are considered valid if they can be successfully parsed. More advanced validation can be provided by seedwing integration.

Encoded uploads will be decoded before checking their validity, and then re-encoded if valid. This does require enough memory to contain the entire SBOM for any `Validator` variant other than `None`.

### CODE REVIEW

File: bombastic/api/src/server.rs [hunk 1]
```diff
@@ -278,13 +279,13 @@ async fn publish_sbom(
     let typ = verify_type(content_type)?;
     let enc = verify_encoding(req.headers().get(CONTENT_ENCODING))?;
     let id = &params.id;
-    let mut payload = payload.map_err(|e| match e {
-        PayloadError::Io(e) => e,
-        _ => io::Error::new(io::ErrorKind::Other, e),
+    let payload = payload.map_err(|e| match e {
+        PayloadError::Io(e) => StorageError::Io(e),
+        _ => StorageError::Io(io::Error::new(io::ErrorKind::Other, e)),
     });
```
Comment (bobmcwhirter) [inline bombastic/api/src/server.rs:285]: Should this have a ? to return the Err?

Comment (jcrossley3) [reply bombastic/api/src/server.rs:285]: I don't think so. It's just mapping of `PayloadError` to `StorageError`

Comment (bobmcwhirter) [reply bombastic/api/src/server.rs:285]: I guess my question is `payload` could be either `Ok(...)` or `Err(...)` at the end of the line, yeah? And then it is `.put_stream(...)` into the storage.  Even if it's an `Result::Err`?

File: bombastic/api/src/server.rs [hunk 2]
```diff
@@ -278,13 +279,13 @@ async fn publish_sbom(
...
     let size = state
         .storage
-        .put_stream(id, typ.as_ref(), enc, &mut payload)
+        .put_stream(id, typ.as_ref(), enc, payload)
```
Comment (bobmcwhirter) [inline bombastic/api/src/server.rs:288]: Here... payload may still be `Err` variant, yeah?

File: bombastic/api/src/server.rs (hunk 1 above)
Comment (bobmcwhirter) [reply bombastic/api/src/server.rs:285]: Ahhh

File: bombastic/api/src/server.rs (hunk 2 above)
Comment (jcrossley3) [reply bombastic/api/src/server.rs:288]: No, it's a virgin stream at this point. `put_stream` will attempt to read from it and if it gets an error, that fn above will be invoked.

**Review by bobmcwhirter — APPROVED** [review]

File: storage/src/validator.rs [hunk 3]
```diff
@@ -77,3 +77,73 @@ async fn check<'a, T, F: Fn(&[u8]) -> Result<T, Error>>(
...
+
+    #[tokio::test]
+    async fn sbom_bzip2_valid() -> Result<(), Error> {
+        let expected = include_bytes!("../../bombastic/testdata/ubi8-valid.json.bz2");
+        let result = test(Validator::SBOM, Some("bzip2"), expected).await?;
+        // This exact file was obtained from an Red Hat internal
+        // repo. I think it's safe to ignore the 4-byte bz2 header, as
+        // it's the block-size (4th byte) that's different: 6 vs 9. I
+        // think we can chalk that up to different bzip2 encoders.
+        Ok(assert_eq!(expected[4..], result[4..]))
```
Comment (helio-frota) [inline storage/src/validator.rs:128]: interesting +1

**Review by lulf — APPROVED** [review]

**Pull request merged**